src/sentry/middleware/integrations/parsers/base.py

```diff
@@ -0,0 +1,143 @@
+from typing import Callable, Iterable, Sequence
+
+from django.http.request import HttpRequest
+from django.urls import ResolverMatch, resolve
+
+from sentry.models.integrations import Integration, OrganizationIntegration
+from sentry.services.hybrid_cloud.organization import ApiOrganizationSummary, organization_service
+from sentry.silo import SiloLimit, SiloMode
+from sentry.silo.client import RegionSiloClient
+from sentry.types.region import Region, get_region_for_organization
+from sentry.utils.sdk import capture_exception
+
+logger = logging.getLogger(__name__)
+
+
+class BaseRequestParser(abc.ABC):
+    """Base Class for Integration Request Parsers"""
+
+    def __init__(self, request: HttpRequest, response_handler: Callable):
+        self.request = request
+        self.match: ResolverMatch = resolve(self.request.path)
+        self.response_handler = response_handler
+        self.error_message = "Integration Request Parsers should only be run on the control silo."
+
+    def _ensure_control_silo(self):
+        if SiloMode.get_current_mode() != SiloMode.CONTROL:
+            logger.error(
+                "request_parser.base.silo_error",
+                extra={"path": self.request.path, "silo": SiloMode.get_current_mode()},
+            )
+            raise SiloLimit.AvailabilityError(self.error_message)
+
+    def get_response_from_control_silo(self):
+        """
+        Used to handle the request directly on the control silo.
+        """
+        self._ensure_control_silo()
+        return self.response_handler(self.request)
+
+    def _get_response_from_region_silo(self, region: Region):
+        region_client = RegionSiloClient(region)
+        return region_client.proxy_request(self.request).to_http_response()
+
+    def get_response_from_region_silos(self, regions: Iterable[Region]):
+        """
+        Used to handle the requests on a given list of regions (synchronously).
+        If multiple regions are provided, only the latest response is returned to the requestor.
+        """
+        self._ensure_control_silo()
+
+        response = None
+
+        with ThreadPoolExecutor(max_workers=len(regions)) as executor:
+            future_to_region = {
+                executor.submit(self._get_response_from_region_silo, region): region
+                for region in regions
+            }
+            for future in as_completed(future_to_region):
+                region = future_to_region[future]
+                try:
+                    region_response = future.result()
+                # This will capture errors from this silo and any 4xx/5xx responses from others
+                except Exception as e:
+                    capture_exception(e)
+                    logger.error(
+                        "request_parser.base.region_proxy_error", extra={"region": region.name}
+                    )
+                else:
+                    response = region_response if region_response is not None else response
```

Comment:
This is a bit odd -- you're collecting just the last response that happened to complete from the futures?  It seems non deterministic and not terribly helpful.

Reply:
I mean I think this leads to the deeper question: What is, the use case for sending the same request to multiple silos and collecting a single response anyways?  Because it's hard to even imagine that clearly.